File: apps/flutter-prototype/test/recovery_coordinator_test.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'package:flutter_test/flutter_test.dart';
import 'package:meshx_flutter_probe/application/recovery_coordinator.dart';
import 'package:meshx_flutter_probe/core/recovery.dart';
import 'package:meshx_flutter_probe/core/recovery_snapshot.dart';

class TestSink implements RecoverySink {
  @override
  Future<void> resume(RecoveryContext context, RecoveryState state) async {
    events.add('resume:${state.cursor}');
  }

  TestSink(this.events);
  final List<String> events;
  Future<void> Function(RecoveryState)? onCommit;
  @override
  Future<void> begin(RecoveryContext context) async {
    events.add('begin');
  }

  @override
  Future<void> snapshot(
    RecoveryContext context,
    List<SnapshotItem> items,
  ) async {
    events.add('snapshot');
  }

  @override
  Future<void> mutations(RecoveryContext context, RecoveryPlan plan) async {
    events.add('apply:${plan.next.cursor}');
  }

  @override
  Future<void> commit(RecoveryContext context, RecoveryState state) async {
    if (onCommit != null) {
      await onCommit!(state);
    } else {
      events.add('commit:${state.cursor}');
    }
  }

  @override
  void publish(RecoveryContext context, RecoveryState state) {
    events.add('publish:${state.cursor}');
  }

  @override
  void quarantine(String reason) {
    events.add(reason);
  }
}

class TestTransport implements RecoveryTransport {
  TestTransport(this.fixture, this.events);
  final Map fixture;
  final List<String> events;
  int cuts = 0, readyCalls = 0, pageSize = 100;
  bool supported = true, resuming = false;
  Future<void> Function()? onRelease;
  Future<Object?> Function(String? token, int limit)? onSnapshot;
  String get epoch => fixture['context']['streamEpoch'];
  @override
  Future<Object?> capabilities() async => {
    'capability': 'meshx.mutation-recovery',
    'versions': supported ? [1] : [],
    'recordVersion': 1,
    'maxPageSize': pageSize,
  };
  @override
  Future<Object?> open(Map<String, dynamic> input, String key) async {
    resuming = input['mode'] == 'resume';
    if (resuming) {
      expect(input['cursor'], {'streamEpoch': epoch, 'position': '2'});
      return {
        'recoveryId': 'resumed',
        'mode': 'resume',
        'streamEpoch': epoch,
        'startCursor': '2',
        'floor': '0',
        'latest': '2',
      };
    }
    return {
      'recoveryId': 'session',
      'mode': 'rebuild',
      'streamEpoch': epoch,
      'startCursor': '0',
      'snapshotBoundary': '0',
      'floor': '0',
      'latest': '0',
      'snapshotId': 'snapshot',
    };
  }

  @override
  Future<Object?> snapshot(String id, String? token, int limit) async =>
      onSnapshot != null
      ? onSnapshot!(token, limit)
      : {
          'snapshotId': 'snapshot',
          'boundary': '0',
          'items': [fixture['directory'], fixture['message']],
          'nextPageToken': null,
          'snapshotComplete': true,
        };
  @override
  Future<Object?> cut(String id) async => {
    'through': resuming ? '2' : '${++cuts}',
    'floor': '0',
    'streamEpoch': epoch,
  };
  @override
  Future<Object?> mutations(
    String id,
    String after,
    String through,
    int limit,
  ) async => {
    'fromExclusive': after,
    'through': through,
    'nextCursor': through,
    'hasMore': false,
    'floor': '0',
    'latest': through,
    'streamEpoch': epoch,
    'records': [
      {
        'recordVersion': 1,
        'eventId': through == '1'
            ? '22222222-2222-4222-8222-222222222222'
            : '33333333-3333-4333-8333-333333333333',
        'streamEpoch': epoch,
        'cursor': through,
        'type': through == '1' ? 'MESSAGE_RECALLED' : 'MESSAGE_UNAVAILABLE',
        'conversationId': 'group:21',
        'messageId': 'message-a',
        'objectVersion': '${int.parse(through) + 1}',
        'committedAt': '2026-09-14T00:00:00.000Z',
      },
    ],
  };
  @override
  Future<Object?> ready(String id, String cursor, bool complete) async {
    readyCalls++;
    events.add('ready:$cursor');
    return {
      'ready': readyCalls > 1,
      'acceptedCursor': cursor,
      'latest': '2',
      'streamEpoch': epoch,
    };
  }

  @override
  Future<void> release(String id) async {
    if (onRelease != null) await onRelease!();
    events.add('release');
  }
}

void main() {
  final f =
      jsonDecode(
            File(
              '../../contracts/test-vectors/recovery-snapshot-pages.json',
            ).readAsStringSync(),
          )
          as Map;
  final owner = RecoveryOwner(
    f['context']['origin'],
    f['context']['userId'],
    1,
  );
  late List<String> events;
  late TestSink sink;
  late TestTransport transport;
  late RecoveryCoordinator runner;
  var current = true;
  setUp(() {
    events = [];
    sink = TestSink(events);
    transport = TestTransport(f, events);
    current = true;
    runner = RecoveryCoordinator(transport, sink, (_) => current);
  });
  test(
    'event-loop cancellation interrupts immediately completed snapshot pages before commit',
    () async {
      var pages = 0;
      transport.pageSize = 1;
      Timer? cancellation;
      transport.onSnapshot = (token, limit) async {
        final index = ++pages;
        cancellation ??= Timer(Duration.zero, runner.cancel);
        return {
          'snapshotId': 'snapshot',
          'boundary': '0',
          'items': [
            if (index == 1)
              {...f['directory'] as Map, 'messageSequenceAtH': '100'},
            if (index > 1)
              {
                ...f['message'] as Map,
                'messageId': 'm$index',
                'messageSequence': '${index - 1}',
              },
          ],
          'nextPageToken': index == 101 ? null : '$index',
          'snapshotComplete': index == 101,
        };
      };
      try {
        expect(await runner.rebuild(owner, 'cancel-paged'), isNull);
        expect(pages, inInclusiveRange(1, 4));
        expect(transport.readyCalls, 0);
        expect(
          events.any(
            (event) =>
                event.startsWith('commit:') || event.startsWith('publish:'),
          ),
          isFalse,
        );
      } finally {
        cancellation?.cancel();
      }
    },
  );
  test(
    'snapshot replay durable commit READY; newer latest needs another commit',
    () async {
      final result = await runner.rebuild(owner, 'request-key');
      expect(result!.phase, RecoveryPhase.onlineSafe);
      expect(runner.durableCursor, '2');
      expect(events, [
        'RECOVERING',
        'begin',
        'snapshot',
        'apply:1',
        'commit:1',
        'ready:1',
        'apply:2',
        'commit:2',
        'ready:2',
        'publish:2',
        'release',
      ]);
    },
  );
  test('pending or failed commit cannot call READY or publish', () async {
    final gate = Completer<void>(), entered = Completer<void>();
    sink.onCommit = (_) async {
      entered.complete();
      await gate.future;
      throw FileSystemException('disk full');
    };
    final running = runner.rebuild(owner, 'request-key');
    await entered.future;
    expect(transport.readyCalls, 0);
    expect(runner.durableCursor, isNull);
    gate.complete();
    expect((await running)!.phase, RecoveryPhase.storageBlocked);
    expect(transport.readyCalls, 0);
    expect(runner.durableCursor, isNull);
    expect(events.any((event) => event.startsWith('publish:')), isFalse);
  });
  test(
    'owner switch during commit cannot publish or release as new owner',
    () async {
      final gate = Completer<void>(), entered = Completer<void>();
      sink.onCommit = (_) async {
        entered.complete();
        await gate.future;
      };
      final running = runner.rebuild(owner, 'request-key');
      await entered.future;
      current = false;
      runner.cancel();
      gate.complete();
      expect(await running, isNull);
      expect(transport.readyCalls, 0);
      expect(events.contains('release'), isFalse);
    },
  );
  test('unsupported capability never opens staging', () async {
    transport.supported = false;
    expect(await runner.rebuild(owner, 'request-key'), isNull);
    expect(runner.reason, 'BLOCKED_UPGRADE');
    expect(events.contains('begin'), isFalse);
  });
  test('cancellation during cleanup cannot return stale ONLINE_SAFE', () async {
    final entered = Completer<void>(), gate = Completer<void>();
    transport.onRelease = () async {
      entered.complete();
      await gate.future;
    };
    final running = runner.rebuild(owner, 'request-key');
    await entered.future;
    current = false;
    runner.cancel();
    gate.complete();
    expect(await running, isNull);
    expect(runner.state!.phase, RecoveryPhase.quarantined);
  });
  test(
    'resume reuses committed data and does not reload snapshot pages',
    () async {
      final first = await runner.rebuild(owner, 'first');
      expect(first?.phase, RecoveryPhase.onlineSafe);
      final count = events.where((e) => e == 'snapshot').length;
      final second = await runner.rebuild(owner, 'resume', resumeFrom: first);
      expect(second?.phase, RecoveryPhase.onlineSafe);
      expect(events.where((e) => e == 'snapshot').length, count);
      expect(events, contains('resume:2'));
      expect(runner.durableCursor, '2');
    },
  );
}

File: apps/web/tests/recovery-coordinator.test.mjs
import assert from 'node:assert/strict'
import {readFileSync} from 'node:fs'
import test from 'node:test'
import {createTsLoader,deferred} from './helpers/load-ts.mjs'
const {RecoveryCoordinator}=createTsLoader()('apps/web/src/services/recoveryCoordinator.ts')
const f=JSON.parse(readFileSync(new URL('../../../contracts/test-vectors/recovery-snapshot-pages.json',import.meta.url)))
const owner={origin:f.context.origin,userId:f.context.userId,generation:1}
function harness() {
  const events=[];let cuts=0,readyCalls=0,current=true
  const sink={begin:async()=>{events.push('begin')},snapshot:async()=>{events.push('snapshot')},mutations:async(_,p)=>{events.push(`apply:${p.next.cursor}`)},
    commit:async(_,s)=>{events.push(`commit:${s.cursor}`)},publish:(_,s)=>events.push(`publish:${s.cursor}`),quarantine:r=>events.push(r)}
  const transport={capabilities:async()=>({capability:'meshx.mutation-recovery',versions:[1],recordVersion:1,maxPageSize:100}),
    open:async()=>({recoveryId:'session',mode:'rebuild',streamEpoch:f.context.streamEpoch,startCursor:'0',snapshotBoundary:'0',floor:'0',latest:'0',snapshotId:'snapshot'}),
    snapshot:async()=>({snapshotId:'snapshot',boundary:'0',items:[f.directory,f.message],nextPageToken:null,snapshotComplete:true}),
    cut:async()=>({through:String(++cuts),floor:'0',streamEpoch:f.context.streamEpoch}),
    mutations:async(_,after,through)=>({fromExclusive:after,through,nextCursor:through,hasMore:false,floor:'0',latest:through,streamEpoch:f.context.streamEpoch,
      records:[{recordVersion:1,eventId:through==='1'?'22222222-2222-4222-8222-222222222222':'33333333-3333-4333-8333-333333333333',streamEpoch:f.context.streamEpoch,
        cursor:through,type:through==='1'?'MESSAGE_RECALLED':'MESSAGE_UNAVAILABLE',conversationId:'group:21',messageId:'message-a',objectVersion:String(Number(through)+1),committedAt:'2026-09-14T00:00:00.000Z'}]}),
    ready:async(_,cursor)=>{readyCalls++;events.push(`ready:${cursor}`);return {ready:readyCalls>1,acceptedCursor:cursor,latest:'2',streamEpoch:f.context.streamEpoch}},release:async()=>{events.push('release')}}
  const runner=new RecoveryCoordinator(transport,sink,()=>current)
  return {runner,sink,transport,events,switchOwner:()=>{current=false},readyCalls:()=>readyCalls}
}
test('snapshot -> replay -> durable commit -> READY; newer latest requires another committed round',async()=>{
  const h=harness(),result=await h.runner.rebuild(owner,'request-key')
  assert.equal(result.phase,'ONLINE_SAFE');assert.equal(h.runner.durableCursor,'2')
  assert.deepEqual(h.events,['RECOVERING','begin','snapshot','apply:1','commit:1','ready:1','apply:2','commit:2','ready:2','publish:2','release'])
})
test('pending or failed commit cannot call READY or publish',async()=>{
  const h=harness(),gate=deferred(),entered=deferred()
  h.sink.commit=async()=>{entered.resolve();await gate.promise;throw new Error('disk full')}
  const running=h.runner.rebuild(owner,'request-key');await entered.promise
  assert.equal(h.readyCalls(),0);assert.equal(h.runner.durableCursor,null)
  gate.resolve();const result=await running
  assert.equal(result.phase,'STORAGE_BLOCKED');assert.equal(h.readyCalls(),0);assert.equal(h.runner.durableCursor,null)
  assert.equal(h.events.some(e=>e.startsWith('publish:')),false)
})
test('owner switch during commit never publishes or sends requests as the new owner',async()=>{
  const h=harness(),gate=deferred(),entered=deferred()
  h.sink.commit=async()=>{entered.resolve();await gate.promise}
  const running=h.runner.rebuild(owner,'request-key');await entered.promise;h.switchOwner();h.runner.cancel();gate.resolve()
  assert.equal(await running,null);assert.equal(h.readyCalls(),0);assert.equal(h.events.includes('release'),false)
})
test('unsupported capability never opens or stages a recovery',async()=>{
  const h=harness();h.transport.capabilities=async()=>({capability:'meshx.mutation-recovery',versions:[],recordVersion:1,maxPageSize:100})
  assert.equal(await h.runner.rebuild(owner,'request-key'),null);assert.equal(h.runner.reason,'BLOCKED_UPGRADE');assert.equal(h.events.includes('begin'),false)
})
test('cancellation during session cleanup cannot return stale ONLINE_SAFE',async()=>{
  const h=harness(),entered=deferred(),gate=deferred()
  h.transport.release=async()=>{entered.resolve();await gate.promise}
  const running=h.runner.rebuild(owner,'request-key');await entered.promise
  h.switchOwner();h.runner.cancel();gate.resolve()
  assert.equal(await running,null);assert.equal(h.runner.state.phase,'QUARANTINED')
})

test('resume keeps committed cursor and skips snapshot download',async()=>{
  const h=harness(),first=await h.runner.rebuild(owner,'first')
  h.sink.resume=async(_,state)=>h.events.push(`resume:${state.cursor}`)
  h.transport.open=async(input)=>{
    assert.equal(input.mode,'resume');assert.equal(input.cursor.position,'2');assert.equal(input.cursor.streamEpoch,f.context.streamEpoch)
    return {recoveryId:'resume',mode:'resume',streamEpoch:f.context.streamEpoch,startCursor:'2',floor:'0',latest:'2'}
  }
  h.transport.cut=async()=>({through:'2',floor:'0',streamEpoch:f.context.streamEpoch})
  h.transport.snapshot=async()=>{throw Error('resume must not download snapshots')}
  const second=await h.runner.rebuild(owner,'resume',first)
  assert.equal(second.phase,'ONLINE_SAFE');assert.equal(h.runner.durableCursor,'2')
  assert.ok(h.events.includes('resume:2'))
})
